Pattern Directory: Merge Core locale, caching updates

iandunn opened the pull request:
## Description

Merges https://core.trac.wordpress.org/changeset/51206 and https://core.trac.wordpress.org/changeset/51208, so the bundled versions of Core files are up to date with Core `5.8-beta4`

Fixes #32754

## How has this been tested?

- [x] Unit tests pass
- [x] Manual testing works

Manual testing:
1. downgrade to wp 5.7 branch so the changesets above don't exist in core
1. point `api.wordpress.org` to a w.org sandbox via `/etc/hosts/` (if applicable)
1. clear the locale transients
1. refresh the editor to trigger a request
2. view the request on the sandbox and see that the payload contains the locale. alternatively, you could observe the request w/ something like Burp Suite, the `pre_http_request` filter, etc.
3.  refresh again and observed that no request is made, because the results from the previous one were cached
4. change locales and refreshe again, and see a new request on the sandbox, and a new transient created locally.




## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

iandunn committed:
Pattern Directory: Send locale, version with remote pattern requests.

This information is needed by the w.org API for translation and caching.

Merges https://core.trac.wordpress.org/changeset/51206
See https://core.trac.wordpress.org/ticket/53435

Co-authored-by: Dion Hulse <dion@wordpress.org>
Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>
Co-authored-by: tellyworth <alex@automattic.com>

iandunn committed:
Pattern Directory: Move caching to endpoint for unique responses.

Now that the pattern API request includes the locale and version, the cache key needs to contain a hash of the query args.

Merges https://core.trac.wordpress.org/changeset/51208
See https://core.trac.wordpress.org/ticket/53435

Co-authored-by: Dominik Schilling <ocean90@git.wordpress.org>
Co-authored-by: Dion Hulse <dion@wordpress.org>
Co-authored-by: Timothy Jacobs <timothyblynjacobs@git.wordpress.org>

iandunn committed:
Pattern Directory: Enable PHPUnit tests.

The tests were failing in `wp-env` because the latest Core updates had not been backported, so they were disabled in #32228. Now that the updates are backported, the tests pass again.

iandunn commented:
The workflow failures are just the e2e tests timing out, they even do that locally. Seems like that's normal lately 🤷🏻‍♂️

Mamaduka commented:
Hi, @iandunn

I think [`should_load_remote_block_patterns`](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/block-patterns.php#L68) is missing from this update.

There were some issues with e2e tests, but I think rebase should fix it.

iandunn committed:
Pattern Directory: Add Core's `should_load_remote_block_patterns` filter.

The filter was added to Core during the process of syncing it with Gutenberg, but wasn't backported to Gutenberg until now.

See https://core.trac.wordpress.org/ticket/53246

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

iandunn commented:
Ah, I only intended to backport the commits I made for https://core.trac.wordpress.org/ticket/53435. It looks like previous backports have rewritten Core code in a different way, and I don't have all the context about the "query" patterns, categories, etc. That makes it more difficult to backport other changes, but I've added the filter in 869e58e8.

I've also rebased to fix the e2e tests.

Beyond that, I'd prefer to keep the scope of this PR limited to the commits I have experience with, and let others backport anything they worked on. 

Related #31922, #32591, #33126

ryelle commented on lib/block-patterns.php:
```diff
@@ -200,15 +200,26 @@ function remove_core_patterns() {
- * Import patterns from wordpress.org/patterns.
+ * Register Core's official patterns from wordpress.org/patterns.
+ *
+ * @since 5.8.0
```
Do we need these `since` tags in Gutenberg? should it reference the GB version instead?

iandunn replied:
I'm not sure, happy to do either.

iandunn replied:
Other files in `lib/` have them sometimes, and it seems to be a roughly 50/50 split for when they reference Core vs Gutenberg. So, this seems fine to me.

ryelle replied:
Okay, works for me 🙂

ryelle reviewed (commented):
Seems to be working as expected on a 5.7 site - tried in English and Japanese, got the correct localized patterns back. Just the one note about `since`.

iandunn commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -85,7 +85,18 @@ public function get_items_permissions_check( $request ) { // phpcs:ignore Variab
+
+		$query_args = array(
+			'locale'     => get_user_locale(),
+			'wp-version' => $wp_version, // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable -- it's defined in `version.php` above.
```
🤔 , i wonder if this should be the _Gutenberg_ version, rather than the _WP_ version? Or maybe we'll want both?

Probably safest to just add the G version too, since we haven't worked out exactly what info the API will need to respond to different compatibility concerns

cc @ryelle , @dd32 , @coreymckrill

ryelle replied:
Is there a way to get the "editor version" in a core-friendly way? This endpoint is for compatibility only, and won't load in 5.8+ because the core class `WP_REST_Pattern_Directory_Controller` exists (unless something changes). So if we add that here, we'll want to add the same arg to core.

I do think the editor version would be a more relevant value for figuring out pattern compatibility, but I don't see the harm in sending both, if editor is possible.

iandunn replied:
That's a good idea. I looked through Core's code and [the docs for the release process](https://make.wordpress.org/core/handbook/about/release-cycle/block-editor-release-process-for-major-releases/), though, and didn't find anything. It's really late in the cycle to be making any changes to Core anyway.

iandunn replied:
I went ahead and added a separate field for the Gutenberg version (and rebased) in b70a3526ead6364831270a01735604b5d266d892
